protoc-gen-go/generator/generator.go

```diff
@@ -1424,32 +1406,43 @@ func (g *Generator) generateImports() {
 	g.P("import " + g.Pkg["proto"] + " " + strconv.Quote(g.ImportPrefix+"github.com/golang/protobuf/proto"))
 	g.P("import " + g.Pkg["fmt"] + ` "fmt"`)
 	g.P("import " + g.Pkg["math"] + ` "math"`)
+	var (
+		imports       = make(map[ImportPath]bool)
+		strongImports = make(map[ImportPath]bool)
+		importPaths   []string
+	)
 	for i, s := range g.file.Dependency {
 		fd := g.fileByName(s)
+		importPath := fd.importPath
 		// Do not import our own package.
-		if fd.PackageName() == g.packageName {
+		if importPath == g.file.importPath {
 			continue
 		}
-		filename := fd.goFileName()
-		// By default, import path is the dirname of the Go filename.
-		importPath := path.Dir(filename)
-		if substitution, ok := g.ImportMap[s]; ok {
-			importPath = substitution
+		if !imports[importPath] {
+			importPaths = append(importPaths, string(importPath))
+		}
+		imports[importPath] = true
+		if !g.weak(int32(i)) {
+			strongImports[importPath] = true
 		}
-		importPath = g.ImportPrefix + importPath
+	}
+	sort.Strings(importPaths)
+	for i := range importPaths {
+		importPath := ImportPath(importPaths[i])
+		packageName := g.PackageName(importPath)
+		fullPath := g.ImportPrefix + string(importPath)
 		// Skip weak imports.
-		if g.weak(int32(i)) {
-			g.P("// skipping weak import ", fd.PackageName(), " ", strconv.Quote(importPath))
+		if !strongImports[importPath] {
+			g.P("// skipping weak import ", packageName, " ", strconv.Quote(fullPath))
 			continue
 		}
 		// We need to import all the dependencies, even if we don't reference them,
 		// because other code and tools depend on having the full transitive closure
 		// of protocol buffer types in the binary.
-		pname := fd.PackageName()
-		if _, ok := g.usedPackages[pname]; !ok {
-			pname = "_"
+		if _, ok := g.usedPackages[importPath]; !ok {
+			packageName = "_"
```

Comment:
I'm not sure I understand from the comment why this is necessary, but whatever.

Reply:
If a `.proto` file imports `a.proto`, we need to import `a.proto`'s Go package for its side effects (i.e., the things it registers) even if we don't reference any of the symbols in it.

In a better world, protos would be IWYU clean and this wouldn't be necessary.

Reply:
That's not exactly what the comment says. Worth updating?

Reply:
I think the comment is fine as is; if you want to update it in a different cl, feel free.